contracts/extensions/VotingReputation.sol

```diff
@@ -0,0 +1,890 @@
+    poll.events[SUBMIT_END] = uint64(now + stakePeriod + submitPeriod);
+    poll.events[REVEAL_END] = uint64(now + stakePeriod + submitPeriod + revealPeriod);
+
+    poll.domainId = _newDomainId;
+    poll.skillId = newDomainSkillId;
+    poll.skillRep = getReputationFromProof(_pollId, address(0x0), _key, _value, _branchMask, _siblings);
+
+    emit PollEscalated(_pollId, msg.sender, _newDomainId);
+  }
+
+  function finalizePoll(uint256 _pollId) public {
+    Poll storage poll = polls[_pollId];
+    require(getPollState(_pollId) == PollState.Finalizable, "voting-rep-poll-not-executable");
+
+    poll.finalized = true;
+
+    bool canExecute = (
+      poll.stakes[NAY] <= poll.stakes[YAY] &&
+      poll.votes[NAY] <= poll.votes[YAY]
+    );
+
+    if (getSig(poll.action) == CHANGE_FUNCTION) {
+      bytes32 structHash = hashExpenditureStruct(poll.action);
+      expenditurePollCounts[structHash] = sub(expenditurePollCounts[structHash], 1);
+
+      // Release the claimDelay if this is the last active poll
+      if (expenditurePollCounts[structHash] == 0) {
+        bytes memory claimDelayAction = createClaimDelayAction(poll.action, 0);
+        executeCall(_pollId, claimDelayAction);
+      }
+
+      uint256 requiredStake = getRequiredStake(_pollId);
+      uint256 votePower = (poll.stakes[NAY] < requiredStake) ? poll.stakes[YAY] : poll.votes[YAY];
+      bytes32 slotHash = hashExpenditureSlot(poll.action);
+
+      if (expenditurePastPolls[slotHash] < votePower) {
+        expenditurePastPolls[slotHash] = votePower;
+        canExecute = canExecute && true;
+      } else {
+        canExecute = canExecute && false;
+      }
+    }
+
+    bool executed;
+
+    if (canExecute) {
+      executed = executeCall(_pollId, poll.action);
+    }
+
+    emit PollFinalized(_pollId, poll.action, executed);
+  }
+
+  /// @notice Claim the staker's reward
+  /// @param _pollId The id of the poll
+  /// @param _permissionDomainId The domain where the extension has the arbitration permission
+  /// @param _childSkillIndex For the domain in which the poll is occurring
+  /// @param _user The user whose reward is being claimed
+  /// @param _vote The side being supported (0 = NAY, 1 = YAY)
+  function claimReward(
+    uint256 _pollId,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    address _user,
+    uint256 _vote
+  )
+    public
+  {
+    Poll storage poll = polls[_pollId];
+    require(
+      getPollState(_pollId) == PollState.Finalized ||
+      getPollState(_pollId) == PollState.Failed,
+      "voting-rep-not-failed-or-finalized"
+    );
+
+    uint256 stake = stakes[_pollId][_user][_vote];
+    uint256 requiredStake = getRequiredStake(_pollId);
+
+    delete stakes[_pollId][_user][_vote];
+
+    uint256 stakerReward;
+    uint256 repPenalty;
+
+    // Went to a vote, use vote to determine reward or penalty
+    if (poll.stakes[NAY] == requiredStake && poll.stakes[YAY] == requiredStake) {
```

Comment:
I'm not sure I'm happy with how this works in conjunction with escalation. If I've escalated, but it didn't get enough backing for a vote once escalated, this logic ends up in the `    // Neither side fully staked, no reward or penalty` branch, which is surely wrong. Ideally, those who staked post-escalation should get what they staked post-escalation back, and the rewards/punishments after the first votes remain as they would have been had the escalation not occurred but I'm not sure how realistic that is.

Reply:
I had toyed around with storing stakes separately per escalation level but that feature never came to maturity. We could change this branch to check for non-zero votes rather than full stakes, which would mean that the most recent vote was used to determine rewards in the case you describe, which might be a step forward.

Reply:
Yeah, that'd be a lot better I think.